lib/core/engine.js

```diff
@@ -160,17 +160,17 @@ class Engine {
 
     this.registerModule('code_generator', {plugins: self.plugins, env: self.env});
 
-    const generateCode = function () {
+    const generateCode = function (modifiedAsset) {
       self.events.request("code-generator:embarkjs:build", () => {
-        self.events.emit('code-generator-ready');
+        self.events.emit('code-generator-ready', modifiedAsset);
       });
     };
-    const cargo = async.cargo((_tasks, callback) => {
-      generateCode();
+    const cargo = async.queue((task, callback) => {
```

Comment:
`generateCode` (which runs the entire pipeline) is executed every time we push to cargo, and we push to cargo when an event is emitted (`'contractsDeployed'`, `'blockchainDisabled'`, or `'asset-changed'`). None of these events emitted pass in more than one object, and so essentially the entire pipeline will be run each time one of those events is emitted. In my opinion, this undercuts the benefits of cargo as it is essentially processing one item at a time. I could be wrong however.

Also, I'm not entirely sure I follow your suggestion of mapping `modifedAsset` to an array of modified assets. Could you post an example, or alternatively, you're welcome to push a commit to the PR.

Reply:
You are right that we call the event to push to cargo for contractDeployed and asset-changed events, however, I added throttle to those events so that if multiple files are saved at once (like on my IDE, which saves everything once you focus out of the IDE), that way, the pipeline event is called only once. 

Therefore, if I was to modify multiple files only once, the behavior is the same for queue and cargo. However, if I were to modify one file, then another, then another, all outside of the throttle window, it would trigger 3 builds for queue, but for the cargo, the first build would trigger, but the two others would get aggregated and sent together.

Finally, for an example:
```
const cargo = async.queue((tasks, callback) => {
  const assets = tasks.map(task => task.modifiedAsset);
  generateCode(assets);
 // In the pipeline, we can then loop the files and check whether we need to run webpack on them
  self.events.once('outputDone', callback);
});
```

Reply:
Good point and makes sense. I've changed back to `async.cargo` and tested the result, and it does work as explained above, very nice!